File: LeetCode/深度优先搜索/101-对称二叉树.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        return self.isDuichen(root, root)

    def isDuichen(self, left, right):
        if not left and not right:
            return True
        if not left or not right:
            return False
        if left.val == right.val:
            return self.isDuichen(left.left, right.right) and self.isDuichen(left.right, right.left)
        return False

File: LeetCode/深度优先搜索/101-对称二叉树.cpp
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution
{
public:
    bool isSymmetric(TreeNode *root)
    {
        return isDuichen(root, root);
    }

    bool isDuichen(TreeNode *left, TreeNode *right)
    {
        if (!left && !right)
        {
            return true;
        }
        if (!left || !right)
        {
            return false;
        }
        if (left->val == right->val)
            return isDuichen(left->left, right->right) && isDuichen(left->right, right->left);
        return false;
    }
};